Framework/Crystal/src/SelectCellOfType.cpp

```diff
@@ -101,44 +101,65 @@ void SelectCellOfType::exec() {
   g_log.notice(std::string(message));
 
   DblMatrix T = info.GetHKL_Tran();
-  g_log.notice() << "Transformation Matrix =  " << T.str() << '\n';
+  g_log.notice() << "Reduced to Conventional Cell Transformation Matrix =  " << T.str() << '\n';
   this->setProperty("TransformationMatrix", T.getVector());
 
-  if (apply) {
-    std::vector<double> sigabc(6);
-    SelectCellWithForm::DetermineErrors(sigabc, newUB, ws, tolerance);
-    //----------------------------------------------
-    o_lattice->setUB(newUB);
+  DblMatrix UBinv = newUB;
+  UBinv.Invert();
 
-    o_lattice->setError(sigabc[0], sigabc[1], sigabc[2], sigabc[3], sigabc[4], sigabc[5]);
+  DblMatrix Tref = UBinv * UB;
+
+  g_log.notice() << "Miller HKL Transformation Matrix =  " << Tref.str() << '\n';
 
+  if (apply) {
+    // Try to optimize(LSQ) to find lattice errors
+    // UB matrix may NOT have been found by unconstrained least squares optimization
+    std::vector<double> sigabc(6);
     int n_peaks = ws->getNumberPeaks();
 
     int num_indexed = 0;
     double average_error = 0.0;
 
-    if (o_lattice->getMaxOrder() == 0) {
-      std::vector<V3D> miller_indices;
-      std::vector<V3D> q_vectors;
-      for (int i = 0; i < n_peaks; i++) {
-        q_vectors.emplace_back(ws->getPeak(i).getQSampleFrame());
-      }
-      num_indexed = IndexingUtils::CalculateMillerIndices(newUB, q_vectors, tolerance, miller_indices, average_error);
-
-      for (int i = 0; i < n_peaks; i++) {
-        IPeak &peak = ws->getPeak(i);
-        peak.setIntHKL(miller_indices[i]);
-        peak.setHKL(miller_indices[i]);
-      }
-    } else {
-      num_indexed = static_cast<int>(num_indexed);
-      for (int i = 0; i < n_peaks; i++) {
-        IPeak &peak = ws->getPeak(i);
-        average_error += (peak.getHKL()).hklError();
-        peak.setIntHKL(T * peak.getIntHKL());
-        peak.setHKL(T * peak.getHKL());
-      }
+    std::vector<V3D> miller_indices;
+    std::vector<V3D> q_vectors;
+
+    V3D modVec1 = Tref * o_lattice->getModVec(0);
+    V3D modVec2 = Tref * o_lattice->getModVec(1);
+    V3D modVec3 = Tref * o_lattice->getModVec(2);
+
+    o_lattice->setUB(newUB);
+
+    o_lattice->setModVec1(modVec1);
+    o_lattice->setModVec2(modVec2);
+    o_lattice->setModVec3(modVec3);
+
+    DblMatrix modHKL(3, 3);
+    modHKL.setColumn(0, modVec1);
+    modHKL.setColumn(1, modVec2);
+    modHKL.setColumn(2, modVec3);
+    o_lattice->setModHKL(modHKL);
+
+    DblMatrix newModUB = newUB * modHKL;
```

Comment:
I must be missing something but it looks like in calculating newModHKL, the `newUB` in this product will cancel out with the `newUB^-1` on the LHS of `Tref`?

Reply:
> A lot of the logic transforming modQ and subtracting modHKL from a satellite peaks (essentially to get another estimate of the parent peak position when determining the errors?) seems to be duplicated across `SelectCellOfType` and `SelectCellWithForm`. Would it be difficult for this to be pulled out and placed in `IndexingUtils` - so both functions could call the same code?

I am fine with that. We could also just put it all in `SelectCellWithForm` like and just call it in `SelectCellOfType`. This was already done with the lattice error calculation.

Reply:
> I must be missing something but it looks like in calculating `newModHKL`, the `newUB` in this product will cancel out with the `newUB^-1` on the LHS of `Tref`?

You are correct, this is not needed. This columns of newModUB should be the same as before as each represents the modulation vector in Q-sample space. I can clean this up to only set the modHKL only. This should make the code clearer.